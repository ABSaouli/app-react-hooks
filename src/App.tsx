import React, { Component, lazy, Suspense } from 'react';
import './App.css';
import { 
  BrowserRouter as Router,
  Route,
 } from 'react-router-dom';
import ErrorBoundary, { FallbackProps } from 'react-error-boundary';
import LoginRoute from './components/LoginRoute';
import Header from './components/Header';
import ProtectedRoute from './components/ProtectedRoute';

// using Lazy to import the component with lazy loading mode
const Home = lazy(() => import('./components/Home'));
const Orders = lazy(() => import('./components/Orders'));
const ParcelEdit = lazy(() => import('./components/Parcels/Edit'));
const OrderEdit = lazy(() => import('./components/Orders/Edit'));

const Login = lazy(() => import('./components/Login'));
function MyFallbackComponent({ componentStack, error }: FallbackProps) {
  return (
    <>
      <h1>Oops! An Error occured!</h1>
      <h3>here is what we know....</h3>
      <p><strong>Error:</strong> {error && error.toString()}</p>
      <p><strong>StackTrace:</strong> {componentStack}</p>
    </>
  )
}

class App extends Component {
  render() {
    return (
      <ErrorBoundary FallbackComponent={MyFallbackComponent} >
        <Router>
          <>
            <Header />
            <Suspense fallback={<h1>Loading ...</h1>}>
                <>
                  <LoginRoute path='/' exact component={Login} />
                  <ProtectedRoute path='/orders' component={Orders} />
                  <ProtectedRoute path='/parcel/edit/:id' component={ParcelEdit} />
                  <ProtectedRoute path='/order/edit/:id' component={OrderEdit} />
                  <ProtectedRoute path='/home' component={Home} />
                </>           
            </Suspense>
          </>
        </Router>
      </ErrorBoundary> 
    );
  }
}

export default App;
