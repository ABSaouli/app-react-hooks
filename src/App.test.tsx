import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow, mount } from 'enzyme';

describe('App Testing', () => {
  it('shallow render <App />', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
})
