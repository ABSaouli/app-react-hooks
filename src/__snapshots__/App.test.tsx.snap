// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Testing shallow render <App /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "FallbackComponent": [Function],
      "children": <BrowserRouter>
        <React.Fragment>
          <withRouter(Header) />
          <UNDEFINED
            fallback={
              <h1>
                Loading ...
              </h1>
            }
          >
            <React.Fragment>
              <LoginRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                exact={true}
                path="/"
              />
              <ProtectedRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/orders"
              />
              <ProtectedRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/parcel/edit/:id"
              />
              <ProtectedRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/order/edit/:id"
              />
              <ProtectedRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/home"
              />
            </React.Fragment>
          </UNDEFINED>
        </React.Fragment>
      </BrowserRouter>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <React.Fragment>
          <withRouter(Header) />
          <UNDEFINED
            fallback={
              <h1>
                Loading ...
              </h1>
            }
          >
            <React.Fragment>
              <LoginRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                exact={true}
                path="/"
              />
              <ProtectedRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/orders"
              />
              <ProtectedRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/parcel/edit/:id"
              />
              <ProtectedRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/order/edit/:id"
              />
              <ProtectedRoute
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/home"
              />
            </React.Fragment>
          </UNDEFINED>
        </React.Fragment>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <withRouter(Header) />,
            <UNDEFINED
              fallback={
                <h1>
                  Loading ...
                </h1>
              }
            >
              <React.Fragment>
                <LoginRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  exact={true}
                  path="/"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/orders"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/parcel/edit/:id"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/order/edit/:id"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/home"
                />
              </React.Fragment>
            </UNDEFINED>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <React.Fragment>
                <LoginRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  exact={true}
                  path="/"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/orders"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/parcel/edit/:id"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/order/edit/:id"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/home"
                />
              </React.Fragment>,
              "fallback": <h1>
                Loading ...
              </h1>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <LoginRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    exact={true}
                    path="/"
                  />,
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/orders"
                  />,
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/parcel/edit/:id"
                  />,
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/order/edit/:id"
                  />,
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/home"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "component": Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    },
                    "exact": true,
                    "path": "/",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "component": Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    },
                    "path": "/orders",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "component": Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    },
                    "path": "/parcel/edit/:id",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "component": Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    },
                    "path": "/order/edit/:id",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "component": Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    },
                    "path": "/home",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": Symbol(react.fragment),
            },
            "type": Symbol(react.suspense),
          },
        ],
        "type": Symbol(react.fragment),
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "FallbackComponent": [Function],
        "children": <BrowserRouter>
          <React.Fragment>
            <withRouter(Header) />
            <UNDEFINED
              fallback={
                <h1>
                  Loading ...
                </h1>
              }
            >
              <React.Fragment>
                <LoginRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  exact={true}
                  path="/"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/orders"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/parcel/edit/:id"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/order/edit/:id"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/home"
                />
              </React.Fragment>
            </UNDEFINED>
          </React.Fragment>
        </BrowserRouter>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <React.Fragment>
            <withRouter(Header) />
            <UNDEFINED
              fallback={
                <h1>
                  Loading ...
                </h1>
              }
            >
              <React.Fragment>
                <LoginRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  exact={true}
                  path="/"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/orders"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/parcel/edit/:id"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/order/edit/:id"
                />
                <ProtectedRoute
                  component={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_ctor": [Function],
                      "_result": null,
                      "_status": -1,
                    }
                  }
                  path="/home"
                />
              </React.Fragment>
            </UNDEFINED>
          </React.Fragment>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <withRouter(Header) />,
              <UNDEFINED
                fallback={
                  <h1>
                    Loading ...
                  </h1>
                }
              >
                <React.Fragment>
                  <LoginRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    exact={true}
                    path="/"
                  />
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/orders"
                  />
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/parcel/edit/:id"
                  />
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/order/edit/:id"
                  />
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/home"
                  />
                </React.Fragment>
              </UNDEFINED>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <React.Fragment>
                  <LoginRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    exact={true}
                    path="/"
                  />
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/orders"
                  />
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/parcel/edit/:id"
                  />
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/order/edit/:id"
                  />
                  <ProtectedRoute
                    component={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      }
                    }
                    path="/home"
                  />
                </React.Fragment>,
                "fallback": <h1>
                  Loading ...
                </h1>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    <LoginRoute
                      component={
                        Object {
                          "$$typeof": Symbol(react.lazy),
                          "_ctor": [Function],
                          "_result": null,
                          "_status": -1,
                        }
                      }
                      exact={true}
                      path="/"
                    />,
                    <ProtectedRoute
                      component={
                        Object {
                          "$$typeof": Symbol(react.lazy),
                          "_ctor": [Function],
                          "_result": null,
                          "_status": -1,
                        }
                      }
                      path="/orders"
                    />,
                    <ProtectedRoute
                      component={
                        Object {
                          "$$typeof": Symbol(react.lazy),
                          "_ctor": [Function],
                          "_result": null,
                          "_status": -1,
                        }
                      }
                      path="/parcel/edit/:id"
                    />,
                    <ProtectedRoute
                      component={
                        Object {
                          "$$typeof": Symbol(react.lazy),
                          "_ctor": [Function],
                          "_result": null,
                          "_status": -1,
                        }
                      }
                      path="/order/edit/:id"
                    />,
                    <ProtectedRoute
                      component={
                        Object {
                          "$$typeof": Symbol(react.lazy),
                          "_ctor": [Function],
                          "_result": null,
                          "_status": -1,
                        }
                      }
                      path="/home"
                    />,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "component": Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      },
                      "exact": true,
                      "path": "/",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "component": Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      },
                      "path": "/orders",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "component": Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      },
                      "path": "/parcel/edit/:id",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "component": Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      },
                      "path": "/order/edit/:id",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "component": Object {
                        "$$typeof": Symbol(react.lazy),
                        "_ctor": [Function],
                        "_result": null,
                        "_status": -1,
                      },
                      "path": "/home",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": Symbol(react.fragment),
              },
              "type": Symbol(react.suspense),
            },
          ],
          "type": Symbol(react.fragment),
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
